build_tag_push.py

```diff
@@ -1,101 +1,155 @@
+        dict_of_contents = yaml.load(file_contents)
+
+    return dict_of_contents
+
+
+def build_check_app(data_from_app_yml, args_app_yml_path,
+                    args_ignore_test_errors):
+    '''
+    This is building and testing their app. Ignoring the return code.
+    '''
+    path_to_build_and_test = args_app_yml_path
+    os.chdir(os.path.dirname(path_to_build_and_test))
+    subprocess.check_call(['bash', data_from_app_yml['build']])
+
+    result = subprocess.call(['bash', data_from_app_yml['test']])
+
+    if result != 0:
+        if args_ignore_test_errors:
+            print("ERROR: Tests failes.")
+        else:
+            Assert.ThrowsException('Unit tests failed.')
+        return False
+
+    else:
+        return True
+
+
+def create_context_directory(context_array, args_app_yml_path):
+    '''
+    Creating a directory that holds the docker contexts that were listed
+    in their app.yml file.
+    '''
+    repo_root_path = os.path.dirname(args_app_yml_path)
+
+    docker_context_path = tempfile.mkdtemp()
+
+    for file_or_dir in context_array:
+        full_path = os.path.join(repo_root_path, file_or_dir)
+        if os.path.isfile(full_path):
+            copyfile(full_path, os.path.join(docker_context_path, file_or_dir))
+        else:
+            shutil.copytree(full_path, os.path.join(docker_context_path,
+                            file_or_dir))
+
+    return docker_context_path
+
+
+def build_docker_image(data_from_app_yml, args_app_yml_path):
+    '''
+    Builds the docker image with the files that are in that directory,
+    which should be a build, static, and Dockerfile. It is then named
+    with their name from their app.yml file.
+    '''
+    docker_dir_context = create_context_directory(data_from_app_yml
+                                                  ['docker_context'],
+                                                  args_app_yml_path)
+
+    image_name = data_from_app_yml['name']
+
+    result = subprocess.check_call(["docker", "build", docker_dir_context,
+                                    "-t", image_name])
+    if result != 0:
+        print("Error: The docker build is not complete.")
+        return False
+    else:
+        return True
+
+
+def tag_docker_image(name_from_app_yml, args_docker_hub_name):
+    '''
+    This function is tagging the Docker Image with the namespace of the
+    user which was passed in as the first argument, and the current git hash.
+    '''
+    subprocess.check_call(["docker", "tag", name_from_app_yml,
+                           os.path.join(args_docker_hub_name,
+                                        name_from_app_yml)])
+
+    current_hash = subprocess.check_output(
+        ["git", "log", "-1", "--pretty=%h"]).strip().decode('ascii')
+
+    result = subprocess.check_call(
+        ["docker", "tag", name_from_app_yml, args_docker_hub_name +
+         '/'+name_from_app_yml + ':'+current_hash])
+
+    if result != 0:
```

Comment:
do we need to check the result? check_call will throw an error if it fails.

Reply:
+1

`subprocess.call` returns the return code and you check it. We use that for unit tests because we don't always want to throw if the test fail. OTOH `check_call` will throw if there's any runtime error so you don't have to check anything.